test/srt/test_triton_int4kv_attention_kernels.py

```diff
@@ -0,0 +1,428 @@
+            k_buffer,
+            k_buffer_int4,
+            k_buffer_scales,
+            quant_group_size,
+        )
+        destindex_copy_quantize_int4kv(
+            v_buffer,
+            v_buffer_int4,
+            v_buffer_scales,
+            quant_group_size,
+        )
+
+        k_extend = torch.empty((extend_token_num, H_KV, D), dtype=dtype, device="cuda")
+        v_extend = torch.empty((extend_token_num, H_KV, D), dtype=dtype, device="cuda")
+        q_extend = torch.empty((extend_token_num, H_Q, D), dtype=dtype, device="cuda")
+        for i in range(B):
+            extend_start_in_buffer = b_start_loc[i] + b_seq_len_prefix[i]
+            extend_end_in_buffer = b_start_loc[i] + b_seq_len[i]
+            extend_start = b_start_loc_extend[i]
+            extend_end = b_start_loc_extend[i] + b_seq_len_extend[i]
+            k_extend[extend_start:extend_end] = k_buffer[
+                extend_start_in_buffer:extend_end_in_buffer
+            ]
+            v_extend[extend_start:extend_end] = v_buffer[
+                extend_start_in_buffer:extend_end_in_buffer
+            ]
+            q_extend[extend_start:extend_end] = torch.empty(
+                (b_seq_len_extend[i], H_Q, D), dtype=dtype, device="cuda"
+            ).normal_(mean=0.1, std=0.2)
+
+        o_extend = torch.empty((extend_token_num, H_Q, D), dtype=dtype, device="cuda")
+        o_redundant = torch.empty(
+            (extend_token_num, H_Q, D), dtype=dtype, device="cuda"
+        )
+
+        b_seq_len_extend = b_seq_len - b_seq_len_prefix
+        b_start_loc_extend = torch.zeros_like(b_seq_len)
+        b_start_loc_extend[1:] = torch.cumsum(b_seq_len_extend[:-1], 0)
+        max_len_extend = torch.max(b_seq_len_extend, 0)[0].item()
+
+        extend_attention_fwd_int4kv(
+            q_extend,
+            k_extend,
+            v_extend,
+            o_extend,
+            k_buffer_int4,
+            v_buffer_int4,
+            k_buffer_scales,
+            v_buffer_scales,
+            req_to_tokens,
+            b_req_idx,
+            b_seq_len,
+            b_seq_len_extend,
+            b_start_loc_extend,
+            max_len_extend,
+            quant_group_size,
+        )
+
+        redundant_attention(
+            q_extend,
+            o_redundant,
+            k_buffer,
+            v_buffer,
+            b_req_idx,
+            b_start_loc,
+            b_seq_len,
+            b_seq_len_prefix,
+            max_len_in_batch,
+        )
+
+        self.assertTrue(torch.allclose(o_extend, o_redundant, atol=5e-2, rtol=5e-2))
+
+    def test_extend_attention(self):
+
+        configs = [
+            (2, 8, 16, 32, 32, 128),
+        ]
+
+        for B, SEQ_PRE, SEQ_EXT, H_Q, H_KV, D in configs:
+            print(
+                f"B, SEQ_PRE, SEQ_EXT, H_Q, H_KV, D: {B} {SEQ_PRE} {SEQ_EXT} {H_Q} {H_KV} {D}"
+            )
+            self._test_extend_attention_once(B, SEQ_PRE, SEQ_EXT, H_Q, H_KV, D)
+
```

Comment:
1. Can you add some unit tests for decode attention? You can compare it with a pytorch implementation.
2. Can you add an end-to-end accuracy test for int8 kv cache? https://github.com/sgl-project/sglang/blob/1d9deeacdb32a5253d7afb25c20d28cb8fc07786/test/srt/test_triton_attn_backend.py#L31

Reply:
it will take time